### PR TITLE
Safe signature

### PR DESCRIPTION
**Pull request by joshaber**
Create a signature using default values if libgit2’s validation fails.

### CODE REVIEW

File: ObjectiveGit/GTRepository.m [hunk 1]
```diff
@@ -718,24 +718,33 @@ - (GTSubmodule *)submoduleWithName:(NSString *)name error:(NSError **)error {
...
+	if (name.length == 0) name = self.class.defaultUserName;
 
 	NSString *email = [configuration stringForKey:@"user.email"];
-	if (email == nil) {
-		NSString *username = NSUserName();
-		if (username.length == 0) username = @"nobody";
-		NSString *domain = NSProcessInfo.processInfo.hostName ?: @"nowhere.local";
-		email = [NSString stringWithFormat:@"%@@%@", username, domain];
-	}
+	if (email == nil) email = self.class.defaultEmail;
```
Comment (jspahrsummers) [inline ObjectiveGit/GTRepository.m]: Why shouldn't this be a length check either?

**Comment by jspahrsummers** [comment]
Can we unart test this too?

File: ObjectiveGit/GTRepository.m (hunk 1 above)
Comment (joshaber) [inline ObjectiveGit/GTRepository.m]: Libgit2 used to be fine with empty emails: https://github.com/libgit2/libgit2/commit/76e3c43fb99eb75cf33affb1a93260af4d472da5

Apparently not anymore :cry:

**Comment by joshaber** [comment]
I went back and forth on whether we should test for `<` and `>` in the email explicitly. That'd let us be more useful, but it'd be duplicating libgit2's test. Thoughts?

**Comment by jspahrsummers** [comment]
> That'd let us be more useful

What do you mean? I was thinking we could just test that invalid strings result in the default.

**Comment by joshaber** [comment]
Ah, sorry, I meant that if we detect that the email address has `<` or `>` in it (which means libgit2 will reject it), we could still try using the configured user name. As is, libgit2 would reject the whole signature and we'd fallback to the default username and email.

**Comment by jspahrsummers** [comment]
If there's a libgit2 API to check validity without actually instantiating a signature, we could test that with the given username and email separately, then both together.

If that sort of thing doesn't exist, I think this implementation is fine.

**Comment by joshaber** [comment]
>If there's a libgit2 API to check validity without actually instantiating a signature, we could test that with the given username and email separately, then both together.

Sadly no.

:fallen_leaf:

**Comment by jspahrsummers** [comment]
:leaves:

**Pull request merged**